### ISSUE TITLE
Metadata configuration sometimes fails to apply on the first Kondo run

### ISSUE THREAD

**Issue by gravieure** [clj-kondo/clj-kondo#2252]
**version** clj-kondo v2023.12.15

**macro usage** Yes.

**platform** macOS/arm64, installed via Homebrew; Linux/amd64 installed via `install-clj-kondo` script, running in Docker.

**editor** n/a

**problem**

My company uses Kondo in a Gitlab pipeline to enforce coding style.  Today, I wrote a very simple macro in one of our projects:

```clojure
(defmacro named-partial [name f & args]
  `(fn ~name [& args#]
     (apply ~f ~@args args#)))
```

Which made Kondo blow up.  Following [the advice of the Kondo documentation](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#metadata-config), I changed this to have metadata:

```clojure
(defmacro named-partial
  {:clj-kondo/ignore [:unresolved-symbol]}
  [name f & args]
  `(fn ~name [& args#]
     (apply ~f ~@args args#)))
```

I pushed this change, and Kondo *still* blew up.  My initial thought was that our CI was using an old version,  since metadata config is a relatively new feature, but I observed the same behavior on v2023.12.15.  After investigating further, I believe this is a Kondo bug.

When I encounter this issue:

1. The first time I lint, errors are reported, even though I've told Kondo to ignore them.
2. Kondo writes a bunch of files to `.clj-kondo`.

If I run the same command again, no errors are reported.

The bug only happens if there's `.clj-kondo` directory present.  If I remove it, no errors are reported and no files are written.  The contents of the directory don't matter, it can be empty, but if it exists, the bug occurs.

The bug only happens when the macro is defined in a different ns than it's used.  If I put the macro in the same ns as it's used, the bug doesn't occur.

**repro**

The problem cannot be reproduced with a single file.  I've created a minimal project which demonstrates the bug, which I've attached to this issue: 
[hanford.tar.gz](https://github.com/clj-kondo/clj-kondo/files/13824140/hanford.tar.gz)

The `README.org` inside it shows how to trigger it, but the tl;dr is:

```shell
$ cd hanford
$ rm -rf .clj-kondo && clj-kondo --lint . # This works fine
$ mkdir .clj-kondo && clj-kondo --lint .  # This fails and leaves a bunch of files in .clj-kondo
$ clj-kondo --lint .                      # After the first failing run, this works.
```

**config**

Not relevant.  The presence of a `.clj-kondo` directory is required to trigger the behavior, but it happens whether there's a `.clj-kondo/config.edn` inside it or not.

**expected behavior**

- clj-kondo should not report errors I have explicitly told it to ignore.
- clj-kondo should not write a bunch of files into `.clj-kondo` when run.

**Comment by borkdude** [comment]
This is how it's supposed to work. When linting code that has metadata configuration, clj-kondo writes this to a local file. Next time you run clj-kondo this configuration goes into effect. This is not a bug.
I recommend running: `clj-kondo --lint "$(clojure -Spath)" --dependencies --copy-configs` before actually linting your code. This will set up clj-kondo will all the necessary information about your dependencies and will also gather all the metadata configuration for subsequent use.

This is also suggested here:

https://github.com/clj-kondo/clj-kondo?tab=readme-ov-file#project-setup

**Comment by gravieure** [comment]
> This is how it's supposed to work. When linting code that has metadata configuration, clj-kondo writes this to a local file. Next time you run clj-kondo this configuration goes into effect. This is not a bug.

It's a documentation bug even if it's not a behavior bug.  [The section on using metadata configs](https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#metadata-config) doesn't mention the requirement that you run Kondo and commit some, but not all, of the files it creates when you do that.

Whether you want to call it a bug or not, it's a horrible experience.

> I recommend running: `clj-kondo --lint "$(clojure -Spath)" --dependencies --copy-configs` before actually linting your code. This will set up clj-kondo will all the necessary information about your dependencies and will also gather all the metadata configuration for subsequent use.

Note that this behavior **is only triggered if I have already done this**, but at a time before the macro has been added.  As  this is in the "project setup" section of the documentation, my expectation is that it's intended to be run once, when the project was set up.  I encountered the problem on a years-old codebase, which I had just added a new macro to.

Is it expected that this command needs to be run every time a macro is added?  If so, is that documented?

**Comment by borkdude** [comment]
> Whether you want to call it a bug or not, it's a horrible experience.

If you don't like the experience, you can configure your macro directly in `.clj-kondo/config.edn`. 
See https://github.com/clj-kondo/clj-kondo/blob/master/doc/config.md#unrecognized-macros for all the options.

> Is it expected that this command needs to be run every time a macro is added? If so, is that documented?

If you have a `.clj-kondo` directory, the necessary configuration will be stored there, after visiting the file with the macro in your editor. If you haven't visited the file with the macro in your editor, then you need to do either that, or execute the above commands. It all boils down to: clj-kondo needs to process the file with the metadata config, before that config is applied to the rest of your project.

I'll take a look at documentation improvements, thanks for the feedback.



I agree that the documentation could be more clear on this.

**Comment by gravieure** [comment]
> It all boils down to: clj-kondo needs to process the file with the metadata config, before that config is applied to the rest of your project.

I note that I have used many other linters for many different languages and formats, and Kondo is the only one which has had this requirement.  It's very surprising behavior.
 
> I'll take a look at documentation improvements, thanks for the feedback.

Some other things that would have made today less unpleasant for me:

- If whatever updates the metadata was a separate command, rather than a side-effect of a different command.
- If the linter gave me better direction (like "update your metadata cache") instead of behaving exactly the same as before I added the metadata.

**Comment by borkdude** [comment]
The issue is that clj-kondo needs to be informed about a macro defined in namespace A while it's linting namespace B, C, D ... and other namespaces that are using namespace A. Clj-kondo does so by saving information from namespace A to the `.clj-kondo` directory, such that when it lints namespace B it doesn't first have to lint namespace A. Can you show me an example of a similar setting for other languages and linters? I'd be curious to learn.

**Comment by gravieure** [comment]
> Can you show me an example of a similar setting for other languages and linters? I'd be curious to learn.

I'm not sure what you mean by "similar."  If you mean, a linter which allows for configuring its errors inline, that's most other linters.  If you want examples, there's [Hadolint](https://github.com/hadolint/hadolint#inline-ignores), [Yamllint](https://yamllint.readthedocs.io/en/stable/disable_with_comments.html), and [Flake8](https://flake8.pycqa.org/en/latest/user/violations.html#in-line-ignoring-errors), all of which I use daily, all of which let you add an inline ignore which takes effect immediately & doesn't require doing anything else.

Given my prior experience and no mention of Kondo's requirements for special handling of inline configs in the docs for that feature, I didn't expect anything else was required.

If the question is "what other linter caches information in the way Kondo does," I don't have an answer.  No other linter has required me to care about its internal implementation details in order to use.

**Comment by borkdude** [comment]
> I'm not sure what you mean by "similar."

What mean with similar is configuration that spans multiple namespaces. You configure the behavior of a function/macro locally in file A but the function/macro is used in file B where you should see the effect.

Since the languages you mention don't support macros, they might not have this problem.

Maybe it's best if you forget about inline configuration for now and just focus on `:config-in-call.`. 

Also I would like you to refer to https://clojure.org/community/etiquette. I sensed a hint of entitlement. This is a free tool that I have poured countless hours of work into. The tone you started this issue with doesn't encourage me to help you any further. Maybe I interpreted it wrong.

**Comment by gravieure** [comment]
I'm sorry if I came off as entitled.  Please understand that Kondo is mandated by my work, I did not choose any of this, and I lost an entire day to debugging this problem.  After spending a lot of time understanding the behavior and writing a detailed report of my problems, I got a dismissive "This is how it's supposed to work" reply after five minutes.

What you're sensing isn't entitlement, but an attempt to moderate an extreme amount of unhappiness and provide some constructive feedback which could prevent others from the deeply negative experience I suffered.

**Comment by borkdude** [comment]
I think I've explained it sufficiently and I'm updating the documentation right now, although it already (briefly) explained how it works. Feel free to create new issues, but please don't take out your frustration on me and keep it to the point.

**Comment by zoren** [comment]
> I'm sorry if I came off as entitled. Please understand that Kondo is mandated by my work, I did not choose any of this, and I lost an entire day to debugging this problem. After spending a lot of time understanding the behavior and writing a detailed report of my problems, I got a dismissive "This is how it's supposed to work" reply after five minutes.
> 
> What you're sensing isn't entitlement, but an attempt to moderate an extreme amount of unhappiness and provide some constructive feedback which could prevent others from repeating the deeply negative experience I suffered.

I think you're taking frustrations from your place of work out on a very productive and generous open source developer. Not a good look in my eyes.
